src/parse_error.rs

```diff
@@ -138,3 +124,195 @@ impl std::fmt::Display for ParseErrorDisplay<'_> {
+            if cursor.goto_next_sibling() {
+                did_visit_children = false;
+            } else if cursor.goto_parent() {
+                did_visit_children = true;
+            } else {
+                break;
+            }
+        } else {
+            if cursor.goto_first_child() {
+                did_visit_children = false;
+            } else {
+                did_visit_children = true;
+            }
+        }
+    }
+    cursor.reset(tree.root_node());
+}
+
+// ------------------------------------------------------------------------------------------------
+// Types to package a tree and parse errors for that tree
+//
+// Parse errors contain `Node` values, that are parametrized by the lifetime `tree of the tree they
+// are part of. It is normally not possible to combine a value and references to that value in a single
+// data type. However, in the case of tree-sitter trees and nodes, the nodes do not point to memory in the
+// tree value, but both point to heap allocated memory. Therefore, moving the tree does not invalidate the
+// node. We use this fact to implement the TreeWithParseError* types.
+//
+// To be able to use these types in errors, we implement Send and Sync. These traits are implemented for
+// Tree, but not for Node. However, since the TreeWithParseError* types contain the tree as well as the nodes,
+// it is okay to implement Send and Sync.
+
+/// A type containing a tree and a parse error
+pub struct TreeWithParseError {
+    tree: Tree,
+    // the 'static lifetime is okay because we own `tree`
+    error: ParseError<'static>,
+}
+
+#[macro_export]
+macro_rules! new_tree_with_parse_error {
+    ($tree:expr,$error:expr) => {{
+        TreeWithParseError {
+            error: unsafe { std::mem::transmute($error) },
+            tree: $tree,
+        }
+    }};
+}
+
+impl TreeWithParseError {
+    pub fn tree(&self) -> &Tree {
+        &self.tree
+    }
+
+    pub fn into_tree(self) -> Tree {
+        self.tree
+    }
+
+    pub fn error(&self) -> &ParseError {
+        &self.error
+    }
+}
+
+impl std::fmt::Debug for TreeWithParseError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{:?}", self.error)
+    }
+}
+
+// Send and Sync must be implemented for ParseError -> Node -> ffi::TSTree
+// This is okay because Send and Sync _are_ implemented for Tree, which also holds ffi::TSTree
+unsafe impl Send for TreeWithParseError {}
+unsafe impl Sync for TreeWithParseError {}
+
+/// A type containing a tree and an optional parse error
+pub struct TreeWithParseErrorOption {
+    tree: Tree,
+    // the 'static lifetime is okay because we own `tree`
+    error: Option<ParseError<'static>>,
+}
+
+#[macro_export]
+macro_rules! new_tree_with_parse_error_option {
+    ($tree:expr,$error:expr) => {{
+        TreeWithParseErrorOption {
+            error: unsafe { std::mem::transmute($error) },
+            tree: $tree,
+        }
+    }};
+}
```

Comment:
I follow your explanation of why it's safe to bundle the `Tree` and its `Node`s and agree with it.  This part gives me pause, though — I think these macros open up too much of a back door on the unsafe code.  Right now anyone* can call the macro and it's not obvious at the call site that unsafe behavior is happening.  Ideally we could write this as a function, so that we could at least verify that `$error` is a valid instead of `ParseError` — but we can't do that, since you'd have to pass in the `Tree` and something containing a reference to that `Tree` as parameters.

_*Not really "anyone", since the macro is filling in some package-local fields, but I think the argument still applies._

Could you instead move the `into_first` method to belong to `TreeWithParseErrorOption`?  Use exactly the same logic to walk the tree and build up the vector, but use this same transmute logic to drop the lifetime.  But you could do it directly, instead of via a call to this macro.

(You could still have the _public_ `into_first` method to live where it currently is, if you want — the proposed one here would then be package-local.)

Reply:
I see how users cannot really use this macro, since it expands at use site and is checked in the context of the expansion 🤦🏻. I think your suggestion is good, to move it into a function of the type itself.

I was looking for a way to give users the possibility to do the same, but I now think that might not be possible in a safe way. Making the fields public certainly is unsound, because it exposes the `Node` with a `static` lifetime.

Reply:
> I was looking for a way to give users the possibility to do the same, but I now think that might not be possible in a safe way.

Yeah I think that's a change that better belongs up in tree-sitter itself — to make `Tree` and `Node` have lifetimes that don't require these kinds of shenanigans in the first place.

> Making the fields public certainly is unsound, because it exposes the `Node` with a `static` lifetime.

Yes good point, I hadn't considered that!  You're right that that also wouldn't work

Reply:
> Yeah I think that's a change that better belongs up in tree-sitter itself - to make Tree and Node have lifetimes that don't require these kinds of shenanigans in the first place.

Agreed. I was wondering though if that is possible. The `Tree` deallocates memory when it is dropped. If you have types `Tree<'a>` and `Node<'a>`, then (a) where would `'a` for the `Tree` come from, and (b) could `'a` be a longer then the lifetime of `Tree<'a>`?